daml-assistant/src/DA/Daml/Assistant/Version.hs

```diff
@@ -260,31 +305,208 @@ instance FromJSON ParsedSdkVersion where
+              case releasedVersionE of
+                Left _ ->
+                  throwIO (CouldNotResolveReleaseVersion targetVersion)
+                Right releasedVersion -> do
+                  _ <- cacheAvailableSdkVersions useCache (\pre -> pure (releasedVersion : fromMaybe [] pre))
+                  pure releasedVersion
+
+resolveSdkVersionToRelease :: UseCache -> SdkVersion -> IO (Either CouldNotResolveVersion ReleaseVersion)
+resolveSdkVersionToRelease _ targetVersion | isHeadVersion targetVersion = pure (Right headReleaseVersion)
+resolveSdkVersionToRelease useCache targetVersion = do
+    resolved <- resolveSdkVersionFromDamlPath (damlPath useCache) targetVersion
+    case resolved of
+      Just resolved -> pure (Right resolved)
+      Nothing -> do
+        let isTargetVersion version =
+              targetVersion == sdkVersionFromReleaseVersion version
+        (releaseVersions, _) <- getAvailableSdkSnapshotVersions useCache
+        case filter isTargetVersion releaseVersions of
+          (x:_) -> pure $ Right x
+          [] -> pure $ Left $ CouldNotResolveSdkVersion targetVersion
+
+resolveReleaseVersionFromDamlPath :: DamlPath -> UnresolvedReleaseVersion -> IO (Maybe ReleaseVersion)
+resolveReleaseVersionFromDamlPath damlPath targetVersion = do
+  let isMatchingVersion releaseVersion =
+          unwrapUnresolvedReleaseVersion targetVersion == releaseVersionFromReleaseVersion releaseVersion
+  resolvedVersions <- getInstalledSdkVersions damlPath
+  pure (find isMatchingVersion resolvedVersions)
+
+resolveSdkVersionFromDamlPath :: DamlPath -> SdkVersion -> IO (Maybe ReleaseVersion)
+resolveSdkVersionFromDamlPath damlPath targetSdkVersion = do
+  let isMatchingVersion releaseVersion =
+          targetSdkVersion == sdkVersionFromReleaseVersion releaseVersion
+  resolvedVersions <- getInstalledSdkVersions damlPath
+  pure (find isMatchingVersion resolvedVersions)
+
+-- | Subset of the github release response that we care about
+data GithubReleaseResponseSubset = GithubReleaseResponseSubset
+  { assetNames :: [T.Text] }
+
+instance FromJSON GithubReleaseResponseSubset where
+  -- Akin to `GithubReleaseResponseSubset . fmap name . assets` but lifted into a parser over json
+  parseJSON = withObject "GithubReleaseResponse" $ \v ->
+    GithubReleaseResponseSubset <$> explicitParseField (listParser (withObject "GithubRelease" (.: "name"))) v "assets"
+
+releaseResponseSubsetSdkVersion :: GithubReleaseResponseSubset -> Maybe T.Text
+releaseResponseSubsetSdkVersion responseSubset =
+  let extractMatchingName :: T.Text -> Maybe T.Text
+      extractMatchingName name = do
+        withoutExt <- T.stripSuffix "-linux.tar.gz" name
+        T.stripPrefix "daml-sdk-" withoutExt
+  in
+  listToMaybe $ mapMaybe extractMatchingName (assetNames responseSubset)
+
+data GithubReleaseError
+  = FailedToFindLinuxSdkInRelease String
+  | Couldn'tParseSdkVersion String InvalidVersion
```

Comment:
You animal
```suggestion
  | CouldNotParseSdkVersion String InvalidVersion
```

Reply:
i refuse to change who i am

Reply:
I beg you not to keep this, it doesn't even save a character lol